baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java

```diff
@@ -0,0 +1,103 @@
+            + "custom executor service. Doing so has very subtle and potentially severe performance implications, so "
+            + "in you're better off using your own executor service which allows you to provide instrumentation "
+            + "and specify the desired parallelism (i.e. the max number of concurrent tasks that will be submitted).\n"
+            + "By default, CompletableFuture uses the globally shared ForkJoinPool. Fork/join pools implement "
+            + "work-stealing, where any thread might steal a task from a different thread's queue when blocked "
+            + "waiting for a subtask to complete. This might not seem like an issue at first glance, but if you rely"
+            + "on the ForkJoinPool for short tasks extensively throughout your codebase and later on you add one "
+            + "piece of code that uses the pool for long (e.g. I/O) tasks, the other parts of your codebase that"
+            + "you'd expect to have consistent performance might experience performance degradation for no apparent "
+            + "reason.\n"
+            + "If you're absolutely certain that the ForkJoinPool is correct, please pass ForkJoinPool.commonPool()"
+            + "directly, ideally with a comment explaining why it is ideal.";
+
+    private static final Matcher<ExpressionTree> SUPPLY_ASYNC = MethodMatchers.staticMethod()
+            .onClass(CompletableFuture.class.getName())
+            .named("supplyAsync")
+            .withParameters(Supplier.class.getName());
+
+    private static final Matcher<ExpressionTree> RUN_ASYNC = MethodMatchers.staticMethod()
+            .onClass(CompletableFuture.class.getName())
+            .named("runAsync")
+            .withParameters(Runnable.class.getName());
+
+    private static final Matcher<ExpressionTree> STATIC_ASYNC_FACTORY_MATCHERS =
+            Matchers.anyOf(SUPPLY_ASYNC, RUN_ASYNC);
+
+    private static final Matcher<ExpressionTree> COMPLETION_STAGE_ASYNC_INVOCATION = MethodMatchers.instanceMethod()
+            .onDescendantOf(CompletionStage.class.getName())
+            .withNameMatching(Pattern.compile(".*Async"));
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        if (STATIC_ASYNC_FACTORY_MATCHERS.matches(tree, state)
+                || isCompletionStageAsyncMethodWithoutExecutor(tree, state)) {
+            return buildDescription(tree).setMessage(ERROR_MESSAGE).build();
+        }
+        return Description.NO_MATCH;
+    }
+
+    private static boolean isCompletionStageAsyncMethodWithoutExecutor(MethodInvocationTree tree, VisitorState state) {
+        if (COMPLETION_STAGE_ASYNC_INVOCATION.matches(tree, state)) {
+            Symbol.MethodSymbol symbol = ASTHelpers.getSymbol(tree);
+            // Should only happen if there are errors in the AST
+            if (symbol != null) {
```

Comment:
Maybe also invert the `if` so it returns early to reduce nesting?

Reply:
👍